### PR TITLE
jbide-23862: fix for spaces in path to oc binary

### PR DESCRIPTION
**Pull request by bdshadow**
@adietish @jeffmaury @robstryker, i've checked it on Fedora 25 and Windows 10.
Please, check it on MacOS and once again on Linux and Windows.

### CODE REVIEW

**Comment by adietish** [comment]
[test]

**Comment by adietish** [comment]
@bdshadow dare to add unit-tests please?
I'm testing it manually...

File: src/main/java/com/openshift/internal/restclient/capability/resources/AbstractOpenShiftBinaryCapability.java [hunk 1]
```diff
@@ -152,9 +154,11 @@ public final void start(final OpenShiftBinaryOption... options) {
 	}
 	
 	private void startProcess(final String location, final OpenShiftBinaryOption... options) {
-		String cmdLine = new StringBuilder(location).append(' ').append(buildArgs(Arrays.asList(options))).toString();
-		String[] args = StringUtils.split(cmdLine, " ");
+		List<String> args = new ArrayList<String>();
+		args.add("oc");
```
Comment (adietish) [inline src/main/java/com/openshift/internal/restclient/capability/resources/AbstractOpenShiftBinaryCapability.java]: dont think that we should hard code the binary name. Can't we just take the last segment of the cmd-line?

**Comment by adietish** [comment]
quite incredible, but even with the fix, which looks fine, things wont work on MacOS:
```
com.openshift.restclient.OpenShiftException: Does your OpenShift binary location exist? Error starting process: Cannot run program "oc" (in directory "/Users/adietish/Downloads/oc binaries/ocp-3.4.1.1.2"): error=2, No such file or directory
	at com.openshift.internal.restclient.capability.resources.AbstractOpenShiftBinaryCapability.startProcess(AbstractOpenShiftBinaryCapability.java:169)
	at com.openshift.internal.restclient.capability.resources.AbstractOpenShiftBinaryCapability.start(AbstractOpenShiftBinaryCapability.java:153)
	at com.openshift.internal.restclient.capability.resources.OpenShiftBinaryPodLogRetrieval$PodLogs.getLogs(OpenShiftBinaryPodLogRetrieval.java:111)
	at com.openshift.internal.restclient.capability.resources.OpenShiftBinaryPodLogRetrieval.getLogs(OpenShiftBinaryPodLogRetrieval.java:67)
	at org.jboss.tools.openshift.internal.ui.job.PodLogsJob$ConsoleStreamPipe.runOCBinary(PodLogsJob.java:162)
	at org.jboss.tools.openshift.internal.core.OCBinaryOperation.run(OCBinaryOperation.java:34)
	at org.jboss.tools.openshift.internal.ui.job.PodLogsJob$ConsoleStreamPipe.run(PodLogsJob.java:152)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.IOException: Cannot run program "oc" (in directory "/Users/adietish/Downloads/oc binaries/ocp-3.4.1.1.2"): error=2, No such file or directory
	at java.lang.ProcessBuilder.start(ProcessBuilder.java:1048)
	at com.openshift.internal.restclient.capability.resources.AbstractOpenShiftBinaryCapability.startProcess(AbstractOpenShiftBinaryCapability.java:165)
	... 7 more
Caused by: java.io.IOException: error=2, No such file or directory
	at java.lang.UNIXProcess.forkAndExec(Native Method)
	at java.lang.UNIXProcess.<init>(UNIXProcess.java:247)
	at java.lang.ProcessImpl.start(ProcessImpl.java:134)
	at java.lang.ProcessBuilder.start(ProcessBuilder.java:1029)
	... 8 more
```

File: src/main/java/com/openshift/internal/restclient/capability/resources/AbstractOpenShiftBinaryCapability.java (hunk 1 above)
Comment (bdshadow) [inline src/main/java/com/openshift/internal/restclient/capability/resources/AbstractOpenShiftBinaryCapability.java]: I think you mean `location`, not `cmdLine`. Yes, i think it's a good proposal. Will change it. Though anyway we hard code such things, for example: https://github.com/openshift/openshift-restclient-java/blob/83f04c2e9c5a7dedef205b5c19faed46de098737/src/main/java/com/openshift/internal/restclient/capability/resources/OpenShiftBinaryRSync.java#L144

**Comment by bdshadow** [comment]
@adietish yes, it's really strange. So now it works everywhere except MacOs

**Comment by adietish** [comment]
@bdshadow great, invoking now works for port forwarding and rsync when launched from JBT.
I have now troubles invoking pod logs which outputs the following:
```
OpenShiftBinaryCapability process exited: error: expected 'logs POD_NAME [CONTAINER_NAME]'.
POD_NAME is a required argument for the logs command
See 'oc logs -h' for help and examples.
```
The arguments that are used are the following:
```
/Users/adietish/.minishift/cache/oc/v3.5.5.8/oc logs --insecure-skip-tls-verify=true --server=https://192.168.64.2:8443  hola-1-1pscn -n hola --token=oFhZZrGozk3KyzZ23-Js0LLMyBtAwHPUSbzcBmfABlU  -f  -c hola
```
which - weird enough - works fine on the cmd-line
```
[/Users/adietish/.minishift/cache/oc/v3.5.5.8/oc, logs, --insecure-skip-tls-verify=true, --server=https://192.168.64.2:8443, , hola-1-1pscn, -n, hola, --token=oFhZZrGozk3KyzZ23-Js0LLMyBtAwHPUSbzcBmfABlU, , -f, , -c, hola]
```

File: src/main/java/com/openshift/internal/restclient/capability/resources/AbstractOpenShiftBinaryCapability.java [hunk 2]
```diff
@@ -152,9 +158,22 @@ public final void start(final OpenShiftBinaryOption... options) {
 	}
 	
 	private void startProcess(final String location, final OpenShiftBinaryOption... options) {
-		String cmdLine = new StringBuilder(location).append(' ').append(buildArgs(Arrays.asList(options))).toString();
-		String[] args = StringUtils.split(cmdLine, " ");
-		ProcessBuilder builder = new ProcessBuilder(args);
+		List<String> args = new ArrayList<String>();
+		ProcessBuilder builder = null;
```
Comment (adietish) [inline src/main/java/com/openshift/internal/restclient/capability/resources/AbstractOpenShiftBinaryCapability.java]: i believe, that for improved readability, we could have l161-l177 extracted to their own method 
```
ProcessBuilder createProcessBuilder(String location, OpenShiftBinaryOption... options)
```

**Comment by bdshadow** [comment]
@adietish a strange comma is here in the middle:
```
--server=https://192.168.64.2:8443, , hola-1-1pscn,
```

**Comment by bdshadow** [comment]
@adietish fixed, check it out, please

**Comment by adietish** [comment]
great!
works for me now (on Mac) when I have the oc binary on a path with spaces. Being tenacious payed out and we won :)
image

**Comment by adietish** [comment]
[merge]

**Comment by adietish** [comment]
[merge]

**Comment by adietish** [comment]
deployed openshift-restclient-java 5.7.0-SNAPSHOT to repository.jboss.org

**Pull request merged**